### PR TITLE
fix: parent_as when used at right side of a combining expression

### PR DESCRIPTION
**Pull request by dgvncsz0f**
This patch fixes a bug in which is impossible to compile a query using
`parent_as` at the right side of a union expression (or any other type
of combination). As an example, consider the following query:

```
  subquery = from(c in Comments, where: parent_as(:posts).id == c.post_id)
  from(p in Posts, where: exists(union_all(subquery, ^subquery)))
```

This query should fail to compile with the following error:

```
** (Ecto.SubQueryError) the following exception happened when compiling a subquery.
    ** (Ecto.QueryError) `parent_as(:post)` can only be used in subqueries in query:
```

The problem, as it seems, is that the `plan_subquery` does not include
the required alias on the combination queries -- the right side query
of the union operator in this case. Thus, when the times come to
resolve the alias, it fails with the above message.

To fix this issue, the function `plan_subquery` has been modified to
include the `@parent_as` alias for every elemen in the `combination`
field. This is done recursevily because the query may have two or more
combining queries.

There is also a required change on ~ecto_adapters~ ecto_sql repository to make
this solution works, which I'll be submiting patches for as well.

### CODE REVIEW

File: test/ecto/query/planner_test.exs [hunk 1]
```diff
@@ -839,6 +839,16 @@ defmodule Ecto.Query.PlannerTest do
     assert Macro.to_string(hd(hd(query.joins).source.query.wheres).expr) =~ "in %Ecto.SubQuery{"
   end
 
+  test "normalize: parent_as used at right side term of union expression" do
```
Comment (dgvncsz0f) [inline test/ecto/query/planner_test.exs:842]: it is not clear to me what kind of tests are necessary to test this change. Please advise.

File: lib/ecto/query/planner.ex [hunk 2]
```diff
@@ -1079,10 +1079,10 @@ defmodule Ecto.Query.Planner do
   end
   defp prewalk_source(%Ecto.SubQuery{query: inner_query} = subquery, kind, query, _expr, counter, adapter) do
     try do
-      inner_query = put_in inner_query.aliases[@parent_as], query
+      inner_query = modify_subquery_for_plan(inner_query, & subquery_put_parent_as_alias(&1, query))
```
Comment (josevalim) [inline lib/ecto/query/planner.ex:1082]:
```suggestion
      inner_query = modify_subquery_for_plan(inner_query, &subquery_put_parent_as_alias(&1, query))
```

**Comment by josevalim** [comment]
Is this still necessary after #3624? I would prefer an approach as in that PR, where we traverse up, rather than changing all combinations as we traverse down, which is bound to be expensive.

**Comment by dgvncsz0f** [comment]
@josevalim I've tested a similar query using the [master branch](https://github.com/elixir-ecto/ecto/commit/6982d4802233cdd8f83d4352edbaee86b7b31090) and also it triggers the error. As the PR you mention is merged, perhaps this is something new?

**Comment by dgvncsz0f** [comment]
> Is this still necessary after #3624? I would prefer an approach as in that PR, where we traverse up, rather than changing all combinations as we traverse down, which is bound to be expensive.

yeah, it makes sense. At the moment. it is not clear to me how to implement this fix the way you are describing, but I can give it another go. If someone more familiar with the codebase could give me some pointers it would be highly appreciated.

**Comment by dgvncsz0f** [comment]
the ecto_sql related changes: https://github.com/elixir-ecto/ecto_sql/pull/322

**Comment by dgvncsz0f** [comment]
👋🏽 any updates on this?

**Comment by josevalim** [comment]
Hi @dgvncsz0f, I would still prefer a mechanism that traverses up, to reduce the cost. So that needs to be investigated, before we can merge this, I haven’t been able to do that yet.

**Comment by josevalim** [comment]
Closing for now. Glad to reopen if the problem still exists after PRs such as: https://github.com/elixir-ecto/ecto_sql/pull/338

**Pull request closed without merging**